peer/abstractlist/list.go

```diff
@@ -0,0 +1,592 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package abstractlist
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"go.uber.org/multierr"
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/introspection"
+	intyarpcerrors "go.uber.org/yarpc/internal/yarpcerrors"
+	"go.uber.org/yarpc/pkg/lifecycle"
+	"go.uber.org/yarpc/yarpcerrors"
+)
+
+var (
+	_noContextDeadlineError = "%q peer list can't wait for peer without a context deadline"
+	_unavailableError       = "%q peer list timed out waiting for peer: %s"
+)
+
+// Implementation is a collection of available peers, with its own
+// subscribers for peer status change notifications.
+// The available peer list encapsulates the logic for selecting from among
+// available peers, whereas a ChooserList is responsible for retaining,
+// releasing, and monitoring peer availability.
+// Use "go.uber.org/yarpc/peer/abstractlist".List in conjunction with an
+// Implementation to produce a "go.uber.org/yarpc/api/peer".List.
+//
+// abstractlist.List and abstractlist.Implementation compose well with sharding
+// schemes the degenerate to returning the only available peer.
+//
+// The abstractlist.List calls Add, Remove, and Choose under a write lock so
+// the implementation is free to perform mutations on its own data without
+// locks.
+type Implementation interface {
+	Add(peer.StatusPeer, peer.Identifier) Subscriber
```

Comment:
The identifier instance is sometimes backed by a concrete type with additional data, data that is not carried by the peer proper. If we were—say—to make a v2, we would have removed the Identifier embed on StatusPeer to eliminate this confusion (and consolidated StatusPeer and Peer).

This is actually the difference between the prior two iterations on this abstract list. We added the separate `peer.Identifier` argument so that it could be carried and cast to `ShardIdentifier`.

Reply:
I don't follow, the `StatusPeer` is also an [interface](https://godoc.org/go.uber.org/yarpc/api/peer#StatusPeer). Are we passing a different concrete type for `StatusPeer` than we do for `peer.Identifier`?

Reply:
They are indeed different concrete types. The peer comes from the transport and the peer identifier comes from the peer list updater. The latter is an address and can carry metadata (carry any of the data from the underlying name resolver *cough*, like ports for other protocols assigned to the same task, or the instance identifier, useful for sharding)